docs/csharp/nullable-attributes.md

```diff
@@ -0,0 +1,301 @@
+
+```csharp
+public void EnsureCapacity<T>([NotNull]ref T[]? storage, int size)
+```
+
+The preceding code expresses the existing contract very clearly: Callers can pass a variable with the `null` value, but the return value is guaranteed to never be null. The `NotNull` attribute is most useful for `ref` and `out` arguments where `null` may be passed as an argument, but that argument is guaranteed to be not null when the method returns. You specify unconditional postconditions using the following attributes:
+
+- [MaybeNull](xref:System.Diagnostics.CodeAnalysis.MaybeNullAttribute): A non-nullable return value may be null.
+- [NotNull](xref:System.Diagnostics.CodeAnalysis.NotNullAttribute): A nullable return value will never be null.
+
+## Specify conditional post-conditions: `NotNullWhen` and `MaybeNullWhen`
+
+You're likely familiar with the `string` method <xref:System.String.IsNullOrEmpty(System.String)?DisplayProperty=nameWithType>. This method returns `true` when the argument isn't null, and not the empty string. It is a form of null-check: Callers don't need to null-check the argument if the method returns `false`. To make a method like this nullable aware, you'd set the argument to a nullable type, and add the `NotNullWhen` attribute:
+
+```csharp
+bool IsNullOrEmpty([NotNullWhen(false)]string? value);
+```
+
+That informs the compiler that any code where the return value is `false` need not be checked. The addition of the attribute informs the compiler's static analysis that `IsNullOrEmpty` performs the necessary null check: It returns `false` when the input argument is not `null`.
+
+```csharp
+string? userInput = GetUserInput();
+if (!(string.IsNullOrEmpty(userInput))
+{
+   int messageLength = userInput.Length; // no null check needed.
+}
+// null check needed on userInput here.
+```
+
+The <xref:System.String.IsNullOrEmpty(System.String)?DisplayProperty=nameWithType> method will be annotated as shown above for .NET Core 3.0. You may have similar methods in your codebase that checks the state of objects for null values. The compiler won't recognize custom null check methods, and you'll need to add the annotations yourself. When you add the attribute, the compiler's static analysis knows when the tested variable has been null checked.
+
+Another use for these attributes is the `Try*` pattern. The postconditions for `ref` and `out` variables are communicated through the return value. Consider this method shown earlier:
+
+```csharp
+bool TryGetMessage(string key, out string message)
+```
+
+The preceding method follows a typical .NET idiom: the return value indicates if `message` was set or not. If the method returns `true`, the value isn't null; `false`, and the method didn't set `message`.
```

Comment:
If the `message` was assigned some value before the call, that value would be cleared with `null` if the method returns `false`, so the `message` is always set:
```suggestion
The preceding method follows a typical .NET idiom: the return value indicates if `message` was set to the found value or, if no message is found, to the default value. If the method returns `true`, the value of `message` isn't null; otherwise, the method sets `message` to null.
```